## Evaluation of Knowledge Graph for RAG

In [1]:
from database.GraphDB import KnowledgeGraphDB
from retrieval.retrieval_hub import RetrievalHub
from query_and_ranking.query import Query
from query_and_ranking.rerank import Reranker
from agentic import Agent
from tqdm import tqdm

2024-06-10 01:45:51.974210: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:9261] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-06-10 01:45:51.974232: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:607] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-06-10 01:45:51.974825: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1515] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-06-10 01:45:52.541908: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [2]:
import os 
from dotenv import load_dotenv
load_dotenv()

neo4j_uri = os.getenv('NEO4J_URI')
neo4j_user = os.getenv('NEO4J_USER')
neo4j_password = os.getenv('NEO4J_PASSWORD')

In [3]:
kg = KnowledgeGraphDB(uri=neo4j_uri, user=neo4j_user, password=neo4j_password)
retrieval_hub = RetrievalHub(kg)
reranker = Reranker()
cv_query = Query(kg, retrieval_hub)

In [4]:
chat_agent = Agent(kg, retrieval_hub, reranker, cv_query, routing='aws',jd_extraction='local')

Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.


Loading checkpoint shards:   0%|          | 0/4 [00:00<?, ?it/s]

In [5]:
import json
questions = json.loads(open('data/benchmark/qa_cv.json').read())

In [6]:
answers = []
i = 0

for question in tqdm(questions):
    prompt = question['prompt']
    answer = dict()
    answer['prompt'] = prompt
    answer['answer'] = question['answer']
    try:
        res = chat_agent.get_cv_from_jd(prompt)
        ans = res['files']
        answer['predict'] = [a.replace('data/raw/data/clean4.0/', '') for a in ans]
    except:
        print('Error')
        answer['predict'] = []
    answers.append(answer)
    with open('data/benchmark/qa_cv_kg.json', 'w') as f:
        f.write(json.dumps(answers))
    i += 1


  0%|          | 0/39 [00:00<?, ?it/s]

Extracting Job Description
End LLM
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['marketing analyst', 'marketing officer', 'marketing manager', 'marketing assistant'], 'suitableList': ['marketing manager', 'marketing executive', 'director of marketing', 'marketing officer', 'head of marketing', 'marketing team leader']}
Extended Num Application:  15


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, business administration - marketing, fpt university, (GPA: 3.2)\nWork:\n- nha khoa parkway (1 years)\n- tổ chức giáo dục fpt (0.5 years)\n- trung tâm tin học ict hà nội (1 years)\n- công ty cổ phân đầu tư và phát triển hạ tầng việt nam (1.5 years)\n- sconnect vietnam (1.5 years)\nRoles:\n- content marketing (1.5 years)\n- digital marketing (1 years)\n- social media manager (1.5 years)\n- marketing executive (1.5 years)\nSkills:\n- market research\n- content creation\n- social media marketing\n- digital marketing\n- advertising campaign management\n- customer journey optimization\n- team management\n- seo\nCities:\n- ha noi\n', metadata={'id': 1749}), -225.46678579783702), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information technology, hano

Not enough application
New Query:  {'cityList': ['ha noi'], 'suitableList': ['marketing staff', 'marketing manager', 'marketing executive', 'director of marketing', 'marketing officer', 'head of marketing', 'marketing', 'marketing team leader']}
[786, 1749, 1802, 2109, 424]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  12
Rerank
[1513, 1593, 1749, 1802, 2006, 2109, 222, 395, 424, 595, 786, 841]
Time: 12.240367412567139
Extracting Job Description


  5%|▌         | 2/39 [00:23<07:08, 11.59s/it]

End LLM
New Query:  {'languageList': ['English'], 'cityList': ['ho chi minh'], 'eduList': ['phd', 'master'], 'roleList': ['marketing analyst', 'marketing officer', 'marketing manager', 'marketing assistant']}
Extended Num Application:  -1
New Query:  {'cityList': ['ho chi minh'], 'roleList': ['marketing manager', 'marketing executive', 'marketing analyst', 'marketing officer', 'marketing', 'marketing assistant']}
Rerank
[468, 92]
Time: 11.133100271224976
Extracting Job Description
End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  10
Reduce CV  172  ->  10  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  10
Rerank
Reranking


  8%|▊         | 3/39 [00:32<06:09, 10.27s/it]

[83, 708, 60, 50, 648, 1953, 2023]
Time: 8.688434362411499
Extracting Job Description
End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['odoo developer'], 'programmingList': ['ros', 'django', 'mongodb', 'opl', 'groovy']}
Extended Num Application:  3
Not enough application
New Query:  {'roleList': ['developer', 'java developer', 'odoo developer', 'python developer', 'unity developer'], 'programmingList': ['ros', 'django', 'mongodb', 'as', 'opl', 'groovy']}
[502, 250, 705, 409, 1953, 1715, 1500]
More -> Reduce CV
Num CV ok:  3
Rerank
Reranking


 10%|█         | 4/39 [00:41<05:51, 10.04s/it]

[1953, 1715]
Time: 9.689563274383545
Extracting Job Description


 13%|█▎        | 5/39 [00:51<05:39,  9.98s/it]

Expecting value: line 15 column 75 (char 482)
invalid syntax (<unknown>, line 15)
Error in converting JSON response
Error
Extracting Job Description


--- Logging error ---
Traceback (most recent call last):
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 1110, in emit
    msg = self.format(record)
          ^^^^^^^^^^^^^^^^^^^
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 953, in format
    return fmt.format(record)
           ^^^^^^^^^^^^^^^^^^
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 687, in format
    record.message = record.getMessage()
                     ^^^^^^^^^^^^^^^^^^^
  File "/home/quanghung20gg/anaconda3/lib/python3.11/logging/__init__.py", line 377, in getMessage
    msg = msg % self.args
          ~~~~^~~~~~~~~~~
TypeError: not all arguments converted during string formatting
Call stack:
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/ipykernel_launcher.py", line 17, in <module>
    app.launc

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['data analyst', 'data scientist', 'research scientist', 'data science / artificial intelligence'], 'role_exp': 0.5, 'suitableList': ['analyst', 'marketing analyst', 'financial market data analyst', 'research analyst', 'finance analyst', 'data scientist', 'data analyst', 'data analysis', 'business analyst', 'sales analyst', 'statistician']}
Extended Num Application:  14
Reduce CV  19  ->  14  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['data scientist'], 'role_exp': 0.5, 'suitableList': ['analyst', 'marketing analyst', 'financial market data analyst', 'finance analyst', 'data scientist', 'data analyst', 'data analysis', 'business analyst', 'sales analyst', 'statistician']}
Risk:  True
Not enough application
Reduce Risk -> Get more CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'dip

 15%|█▌        | 6/39 [01:01<05:26,  9.90s/it]

[606, 1456, 280, 270, 1473, 1291, 1690, 1847, 2141, 1428, 135]
Time: 9.745292901992798
Extracting Job Description


 18%|█▊        | 7/39 [01:11<05:19, 10.00s/it]

End LLM
New Query:  {'institutionList': ['hanoi university of science and technology', 'hanoi university of industry'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  7
Reduce CV  11  ->  7  CV
New Query:  {'institutionList': ['hanoi university of industry'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Risk:  True
Num CV ok:  5
Rerank
[1802, 1896, 2004, 651, 658]
Time: 10.198487520217896
Extracting Job Description


 21%|██        | 8/39 [01:20<04:56,  9.55s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['accountant', 'al audit ', 'al audit manager', 'auditor']}
Extended Num Application:  4
Reduce CV  7  ->  4  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['accountant', 'al audit ', 'auditor']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  4
Rerank
Reranking
[1817, 138, 738]
Time: 8.592268705368042
Extracting Job Description


 23%|██▎       | 9/39 [01:28<04:31,  9.05s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  3
Reduce CV  172  ->  3  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  3
Rerank
Reranking
[60, 1953]
Time: 7.95912504196167
Extracting Job Description


 26%|██▌       | 10/39 [01:36<04:15,  8.81s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  1
Reduce CV  172  ->  1  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  1
Rerank
[1953]
Time: 8.252793073654175
Extracting Job Description
End LLM
New Query:  {'institutionList': ['national economics university (neu)'], 'roleList': ['analyst', 'data science / artificial intelligence', 'financial market data analyst', 'data scientist', 'data analyst', 'data analysis', 'research scientist'], 'suitableList': ['machine learning engineer', 'analyst', 'data science / artificial intelligence', 'financial market data analyst', 'research analyst', 'data scientist', 'data analyst', 'data analysis', 'research scientist']}
Extended Num Application:  12
Not enough application
New Query:  {'institutionList': ['national economics university (neu)', 'university of econom

 28%|██▊       | 11/39 [01:48<04:39,  9.97s/it]

[1214, 124, 1291, 1690, 1941, 2013, 732, 606, 561, 280, 463, 495]
Time: 12.62506365776062
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, computer programming, fpt polytechnic hà nội, (GPA: 3.6)\nWork:\n- công ty cổ phần thương mại và phát triển công nghệ bách việt (1 years)\nRoles:\n- web developer (1 years)\n- it developer (1 years)\nSkills:\n- system design\n- system analysis\n- proxy deployment\n- tool development\nProgramming languages:\n- python (1 years)\n- php (1 years)\n- javascript (1 years)\nCities:\n- ha noi\n', metadata={'id': 2006}), -238.15044894781138), (Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, web development, fpt polytechnic college, (GPA: 3.6)\nRoles:\n- web developer (0 years)\nSkills:\n- web development\n- project management\n- teamwork\n- problem solving\n- communication\nProgramming languages:\n- php (0 years)\n- mysql (0 years)\n- html (0 year

End LLM
New Query:  {'cityList': ['ha noi'], 'institutionList': ['fpt polytechnic college', 'fpt university hcm', 'fpt university'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  14
Not enough application
New Query:  {'cityList': ['ha noi'], 'institutionList': ['fpt polytechnic college', 'fpt university hcm', 'fpt university']}
[862, 776, 648, 1703, 1749, 64, 279, 358]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  11
Rerank
[1593, 1703, 1749, 2006, 279, 358, 395, 64, 648, 776, 862]
Time: 10.672229528427124
Extracting Job Description


 33%|███▎      | 13/39 [02:08<04:15,  9.83s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['digital marketing manager', 'marketing officer', 'marketing manager', 'content marketing manager'], 'suitableList': ['marketing manager', 'marketing executive', 'brand manager', 'director of marketing', 'marketing officer', 'head of marketing', 'digital marketing manager', 'marketing team leader', 'content marketing manager']}
Extended Num Application:  4
Num CV ok:  4
Rerank
Reranking
[972, 468, 1139]
Time: 9.015899658203125
Extracting Job Description
End LLM
New Query:  {'institutionList': ['national economics university (neu)'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['dataops engineer', 'data analyst', 'machine learning engineer', 'data scientist'], 'suitableList': ['machine learning engineer', 'financial market data analyst', 'dataops engineer', 'data analyst', 'data scientist', 'data analysis']}
Extended 

 36%|███▌      | 14/39 [02:17<03:58,  9.53s/it]

[1690, 280, 561]
Time: 8.824827671051025
Extracting Job Description


 38%|███▊      | 15/39 [02:26<03:41,  9.24s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['back-end developer', 'front end developer', 'full stack developer', 'developer'], 'role_exp': 2.0}
Extended Num Application:  3
Reduce CV  5  ->  3  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['back-end developer', 'front end developer'], 'role_exp': 2.0}
Risk:  True
Num CV ok:  3
Rerank
Reranking
[2083, 1579]
Time: 8.569492101669312
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information security, ho chi minh city university of technology and food industry, (GPA: 2.9)\nWork:\n- công ty c ổ  ph ầ n công ngh ệ tam khoadeveloper .net (1 years)\nRoles:\n- tester (0 years)\n- researcher (0.5 years)\n- software developer (1 years)\nSkills:\n- software development\n- machine learning\n- sql\n- technical support\nProgramming languages:\n- .net (1 years)\n- python (0.5 years)\n- c# (1 years)\n- sql (1 years)\nCities:\n- ho chi minh\n', metadata={'id': 1841}), -191.59035470182258), (Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, information technology, ho chi minh city open university, (GPA: 3.6)\nWork:\n- m q -h o use (1.5 years)\n- healthy heart project (0.5 years)\nRoles:\n- mobile app developer (0.5 years)\n- web develo

End LLM
New Query:  {'cityList': ['ho chi minh'], 'institutionList': ['ho chi minh city university of industry and trade', 'ho chi minh city industry and trade college'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  6
Not enough application
New Query:  {'cityList': ['ho chi minh'], 'institutionList': ['ho chi minh city university of industry and trade', 'ho chi minh city industry and trade college']}
[178, 191]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  4
Rerank
[1240, 178, 1841, 191]
Time: 9.361817121505737
Extracting Job Description


 44%|████▎     | 17/39 [02:44<03:23,  9.26s/it]

End LLM
New Query:  {'institutionList': ['ho chi minh city open university'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  7
Num CV ok:  6
Rerank
Reranking
[548, 1240, 1669, 1604, 1193]
Time: 9.234389543533325
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\n- Vietnamese\nMajors:\n- bachelor, computer science and data science, ho chi minh city university of science, (GPA: 3.6)\nWork:\n- fundiininc. (paylater) (0.5 years)\nRoles:\n- data analyst (0.5 years)\n- research engineer (0.5 years)\nSkills:\n- data analysis\n- machine learning\n- cloud computing\n- natural language processing\nProgramming languages:\n- python (0.5 years)\n- sql (0.5 years)\n- c++ (0.5 years)\nCities:\n- ho chi minh\nCertifications:\n- SQL (Basic)\n- Problem Solving (Intermediate)\n- Mathematics for Machine Learning: Linear Algebra\n- The Python Creative Programming Course\nAwards:\n- 2nd Place, Classification of plants of Southeast Asia - ISODS Competitive Online\n', metadata={'id': 1428}), -288.328682225416), (Document(page_content='Languages:\n- English\nMajors:\n- 

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['ba'], 'role_exp': 0.5, 'suitableList': ['analyst', 'marketing analyst', 'financial market data analyst', 'research analyst', 'finance analyst', 'data scientist', 'data analyst', 'data analysis', 'business analyst', 'sales analyst']}
Extended Num Application:  7
Not enough application
New Query:  {'roleList': ['developer', 'admin', 'ba'], 'role_exp': 0.45, 'suitableList': ['analyst', 'marketing analyst', 'financial market data analyst', 'research analyst', 'finance analyst', 'data scientist', 'data analyst', 'data analysis', 'business analyst', 'sales analyst', 'valuation analyst']}
[]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  5
Rerank
[135, 1428, 1473, 443, 548]
Time: 8.704793691635132
Extracting Job Description


 49%|████▊     | 19/39 [03:01<02:58,  8.94s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['embedded developer', 'it developer', 'embedded systems engineer', 'developer']}
Extended Num Application:  3
Reduce CV  18  ->  3  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['embedded developer']}
Risk:  True
Not enough application
Reduce Risk -> Get more CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['embedded developer', 'embedded systems engineer', 'developer']}
[850, 642, 587, 409, 345, 92, 23, 1953, 1493, 1500, 1669, 705, 2073, 279, 83, 1715, 1703]
More -> Reduce CV
Num CV ok:  3
Rerank
Reranking
[642, 1953]
Time: 8.56669569015503
Extracting Job Description


 51%|█████▏    | 20/39 [03:12<03:00,  9.50s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['it support']}
Extended Num Application:  3
Num CV ok:  2
Rerank
[1530, 178]
Time: 10.8257737159729
Extracting Job Description


 54%|█████▍    | 21/39 [03:22<02:52,  9.59s/it]

End LLM
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['developer', ' web developer', 'mobile app developer', 'android developer'], 'suitableList': ['software developer', 'flutter developer', 'developer', ' web developer', 'android developer', 'mobile app developer', 'java developer', 'unity developer', 'it developer']}
Extended Num Application:  -1
New Query:  {'cityList': ['ha noi'], 'suitableList': ['software developer', 'flutter developer', 'developer', ' web developer', 'android developer', 'game developer', 'mobile app developer', 'java developer', 'unity developer', 'embedded developer', 'it developer']}
New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['developer', ' web developer', 'mobile app developer', 'android developer'], 'suitableList': ['software developer', 'flutter developer', 'developer', ' web de

 56%|█████▋    | 22/39 [03:31<02:39,  9.40s/it]

Time: 8.964489459991455
Extracting Job Description
End LLM
New Query:  {'cityList': ['ho chi minh'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['back-end developer', 'front end developer', 'full stack developer', 'developer'], 'role_exp': 0.0}
Extended Num Application:  7
Reduce CV  20  ->  7  CV
New Query:  {'cityList': ['ho chi minh'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['back-end developer', 'front end developer', 'full stack developer', 'developer'], 'role_exp': 0.0}
Still more -> Reduce CV
Risk:  True
Num CV ok:  7
Rerank
Reranking


 59%|█████▉    | 23/39 [03:41<02:31,  9.48s/it]

[751, 1021, 595, 439, 37]
Time: 9.661569595336914
Extracting Job Description
End LLM
New Query:  {'eduList': ['phd', 'master'], 'roleList': ['developer', 'front end developer', 'full stack developer', 'back-end developer'], 'role_exp': 5.0, 'skillList': ['technical support', 'employee relations', 'administrative support', 'employer br', 'role', 'system implementation']}
Extended Num Application:  -1
New Query:  {'roleList': ['back-end developer', 'full stack developer', 'developer', 'front end developer', 'game developer', 'it developer'], 'role_exp': 4.25, 'skillList': ['technical support', 'employee relations', 'administrative support', 'network administration', 'employer br', 'role', 'flexibility', 'organization', 'system implementation']}
Rerank
[1000, 1021, 104, 1041, 1066, 1092, 1104, 1127, 1139, 1165, 1193, 1214, 124, 1240, 1253, 1274, 1291, 1308, 1313, 1322, 1348, 135, 1371, 138, 1388, 1396, 1410, 1428, 1438, 1456, 1462, 1473, 1488, 1493, 1500, 1513, 1530, 1553, 1560, 1579, 158

 62%|██████▏   | 24/39 [03:51<02:25,  9.67s/it]

Time: 10.132424354553223
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- undergraduate, software development, ispace cyber security college, (GPA: 3.6)\nRoles:\n- backend developer (0 years)\nSkills:\n- role\n- based access control\n- authentication\n- docker\n- clean architecture\n- api development\n- database management\n- middleware\n- jwt\nProgramming languages:\n- c# (0 years)\n- javascript (0 years)\n- asp .net core (0 years)\n- react (0 years)\n- node (0 years)\n- express (0 years)\n- mysql (0 years)\n- mongodb (0 years)\nCertifications:\n- AWS Academy Cloud Foundations\nAwards:\n- AWS Academy Cloud Foundations\n', metadata={'id': 50}), -312.8460475299614), (Document(page_content='Languages:\n- English\nWork:\n- dong tien service corporation (5 years)\n- phu nhuan jewelry (1 years)\n- fe credit (2 years)\n- fe credit (2.5 years)\n- sieuviet 

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['dataops engineer', 'devops', 'developer'], 'role_exp': 1.0}
Extended Num Application:  4
Reduce CV  11  ->  4  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['devops'], 'role_exp': 1.0}
Risk:  True
Not enough application
Reduce Risk -> Get more CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['dataops engineer', 'devops'], 'role_exp': 1.0}
[573, 502]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  3
Rerank
[50, 502, 573]
Time: 9.07765245437622
Extracting Job Description


 67%|██████▋   | 26/39 [04:10<02:04,  9.59s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['php developer', 'back-end developer', 'full stack developer', ' web developer', 'front end developer', 'java developer'], 'role_exp': 2.0, 'suitableList': ['php developer', 'back-end developer', 'full stack developer', 'it developer', 'developer', ' web developer', 'front end developer', 'net developer', 'java developer', 'android developer']}
Extended Num Application:  4
Reduce CV  7  ->  4  CV
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['php developer', 'back-end developer', 'full stack developer', ' web developer', 'front end developer', 'java developer'], 'role_exp': 2.0, 'suitableList': ['php developer', 'back-end developer', 'full stack developer', 'developer', ' web developer', 'front end developer', 'net developer', 'java developer']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  4
Rerank
Rer

 69%|██████▉   | 27/39 [04:21<02:00, 10.04s/it]

New Query:  {'cityList': ['ha noi'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['production planner', 'content creator', 'brand marketing', 'marketing', 'strategic planning'], 'role_exp': 1.0, 'suitableList': ['production planner', 'content creator', 'content writer', 'marketing executive', 'marketing communications', 'training ', 'brand marketing', 'marketing', 'content marketing', 'financial planner', 'strategic planning']}
Risk:  True
Num CV ok:  4
Rerank
[1371, 1802, 2109, 424]
Time: 11.07665729522705
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\n- Vietnamese\nMajors:\n- undergraduate, computer networking and communication, ho chi minh city university of technology, (GPA: 3.6)\nWork:\n- hung phat usa (0.5 years)\n- private supermarket (0.5 years)\n- mwg (0.5 years)\n- etomil milk (0.5 years)\n- ho chi minh city university of technology (6.5 years)\nRoles:\n- it support (0.5 years)\n- inventory management staff (0.5 years)\n- security camera supervisor (0.5 years)\n- marketing staff (0.5 years)\n- programmer (0.5 years)\nSkills:\n- system configuration\n- management\n- network administration\n- project management\n- web development\n- mobile development\n- game development\n- security\n- iot\nProgramming languages:\n- html (0.5 years)\n- css (0.5 years)\n- unity (0.5 years)\n- wordpress (0.5 years)\n- sql (0.5 years)\nCities:\n- t

End LLM
New Query:  {'cityList': ['ho chi minh'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['production planner', 'content creator', 'content writer', 'marketing communications', 'training ', 'brand marketing', 'marketing', 'content marketing', 'financial planner', 'strategic planning'], 'role_exp': 1.0, 'suitableList': ['production planner', 'content creator', 'content writer', 'marketing executive', 'marketing communications', 'solution architect', 'project management', 'training ', 'brand marketing', 'social media marketing', 'marketing', 'content marketing', 'financial planner', 'strategic planning', 'content marketing manager']}
Extended Num Application:  4
Not enough application
New Query:  {'cityList': ['ho chi minh'], 'roleList': ['production planner', 'content creator', 'content writer', 'marketing executive', 'marketing communications', 'project management', 'training ', 'brand marketing', 'marketing', 'content marketing

 74%|███████▍  | 29/39 [04:41<01:38,  9.81s/it]

End LLM
New Query:  {'institutionList': ['national economics university (neu)'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  30
Not enough application
New Query:  {'institutionList': ['national economics university (neu)', 'university of economics']}
[1214, 1274, 1690, 424, 463, 8, 1941, 2013, 495, 324, 227, 1783, 310, 561, 732, 406, 606, 280, 738, 124, 1761, 1993, 2109, 1722, 1291]
Num CV ok:  25
Rerank
[1214, 124, 1274, 1291, 1690, 1722, 1761, 1783, 1941, 1993, 2013, 2109, 227, 280, 310, 324, 406, 424, 463, 495, 561, 606, 732, 738, 8]
Time: 8.256222248077393
Extracting Job Description


 77%|███████▋  | 30/39 [04:50<01:26,  9.60s/it]

End LLM
New Query:  {'institutionList': ['university of economics and law', 'university of economics'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  4
Num CV ok:  4
Rerank
Reranking
[692, 443, 1901]
Time: 9.116118907928467
Extracting Job Description


 79%|███████▉  | 31/39 [04:59<01:15,  9.46s/it]

End LLM
New Query:  {'institutionList': ['hcmc open university', 'ho chi minh city open university'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  11
Not enough application
New Query:  {'institutionList': ['hcmc open university', 'ho chi minh city open university', 'international university - ho chi minh city national university']}
[456, 618, 533, 548, 1669, 1604, 1240, 1193]
Num CV ok:  8
Rerank
[1193, 1240, 1604, 1669, 456, 533, 548, 618]
Time: 9.138103008270264
Extracting Job Description


 82%|████████▏ | 32/39 [05:09<01:08,  9.78s/it]

End LLM
New Query:  {'institutionList': ['ho chi minh city university of technology', 'ho chi minh city university of technology and education'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  4
Reduce CV  6  ->  4  CV
New Query:  {'institutionList': ['ho chi minh city university of technology', 'ho chi minh city university of technology and education'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Still more -> Reduce CV
Risk:  True
Num CV ok:  4
Rerank
Reranking
[2027, 2103, 262]
Time: 10.521638870239258
Extracting Job Description


 85%|████████▍ | 33/39 [05:19<00:58,  9.69s/it]

End LLM
New Query:  {'institutionList': ['university of economics and law', 'university of economics'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  3
Reduce CV  4  ->  3  CV
New Query:  {'institutionList': ['university of economics and law'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Risk:  True
Num CV ok:  3
Rerank
Reranking
[443, 1901]
Time: 9.481304168701172
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information security, ho chi minh city university of technology and food industry, (GPA: 2.9)\nWork:\n- công ty c ổ  ph ầ n công ngh ệ tam khoadeveloper .net (1 years)\nRoles:\n- tester (0 years)\n- researcher (0.5 years)\n- software developer (1 years)\nSkills:\n- software development\n- machine learning\n- sql\n- technical support\nProgramming languages:\n- .net (1 years)\n- python (0.5 years)\n- c# (1 years)\n- sql (1 years)\nCities:\n- ho chi minh\n', metadata={'id': 1841}), -222.68203494644047), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information technology, vnu-hcm university of science, (GPA: 3.6)\nRoles:\n- software engineer (0.5 years)\n- web developer (0.5 years)\nSkills:\n- cloud storage\n- devops\n- machine learning\n- database

End LLM
New Query:  {'institutionList': ['university of foreign languages - information technology', 'ho chi minh city university of foreign languages and information technology'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  4
Not enough application
New Query:  {'institutionList': ['university of foreign languages - information technology', 'ho chi minh city university of foreign languages and information technology']}
[409, 555]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  3
Rerank
[1841, 409, 555]
Time: 9.567428588867188
Extracting Job Description


 90%|████████▉ | 35/39 [05:37<00:37,  9.42s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['finance analyst', 'business analyst', 'sales analyst', 'analyst']}
Extended Num Application:  6
Num CV ok:  6
Rerank
Reranking
[548, 443, 1193, 138]
Time: 8.858824253082275
Extracting Job Description


 92%|█████████▏| 36/39 [05:46<00:27,  9.08s/it]

End LLM
New Query:  {'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate'], 'roleList': ['investment consultant', 'consultant', 'financial consultant', 'legal consultant']}
Extended Num Application:  6
Not enough application
New Query:  {'roleList': ['implementation consultant', 'financial consultant', 'legal consultant', 'consultant', 'investment consultant']}
[669, 548, 1193, 456, 1783]
Num CV ok:  5
Rerank
Reranking
[1193, 1783, 456, 548, 669]
Time: 8.286395072937012
Extracting Job Description


 95%|█████████▍| 37/39 [05:55<00:18,  9.17s/it]

End LLM
New Query:  {'institutionList': ['ho chi minh city university of technology and food industry', 'ho chi minh city university of industry and trade'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  3
Num CV ok:  3
Rerank
Reranking
[1841, 1066]
Time: 9.401374578475952
Extracting Job Description


/home/quanghung20gg/anaconda3/lib/python3.11/site-packages/langchain_core/vectorstores.py:330: UserWarning: Relevance scores must be between 0 and 1, got [(Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information security, ho chi minh city university of technology and food industry, (GPA: 2.9)\nWork:\n- công ty c ổ  ph ầ n công ngh ệ tam khoadeveloper .net (1 years)\nRoles:\n- tester (0 years)\n- researcher (0.5 years)\n- software developer (1 years)\nSkills:\n- software development\n- machine learning\n- sql\n- technical support\nProgramming languages:\n- .net (1 years)\n- python (0.5 years)\n- c# (1 years)\n- sql (1 years)\nCities:\n- ho chi minh\n', metadata={'id': 1841}), -216.83886600124606), (Document(page_content='Languages:\n- English\nMajors:\n- bachelor, information technology, vnu-hcm university of science, (GPA: 3.6)\nRoles:\n- software engineer (0.5 years)\n- web developer (0.5 years)\nSkills:\n- cloud storage\n- devops\n- machine learning\n- database

End LLM
New Query:  {'institutionList': ['national economics university hanoi', 'ho chi minh city university of economics'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  11
Not enough application
New Query:  {'institutionList': ['national economics university hanoi', 'ho chi minh city university of economics']}
[1593, 669, 468, 135, 1958]
Less -> Sematic Search
Using Sematic Search
Num CV ok:  8
Rerank
[1240, 135, 1488, 1593, 1841, 1958, 468, 669]
Time: 10.017922163009644
Extracting Job Description
End LLM
New Query:  {'institutionList': ['hanoi university of science and technology', 'hanoi university of business and technology', 'university of technology information - vnu', 'posts and telecommunications institute of technology', 'vnu-hcm university of science'], 'eduList': ['master', 'bachelor', 'undergraduate', 'diploma', 'phd', 'certificate']}
Extended Num Application:  12
Num CV ok:  10
Rerank
Reranking


100%|██████████| 39/39 [06:15<00:00,  9.64s/it]

[751, 345, 222, 2083, 1104, 19, 1733, 1488, 2051]
Time: 10.418983221054077


In [10]:
recall = 0
precision = 0
long_recall = 0
long_precision = 0
len_long = 0
short_recall = 0
short_precision = 0
len_short = 0

for answer in answers:
    num_ans = len(answer['answer'])
    correct = 0
    ans = set(answer['answer'])
    for p in answer['predict']:
        if p in ans:
            correct += 1
    recall += correct / num_ans
    precision += correct / (len(answer['predict'])+1e-5)
    if num_ans > 7:
        long_recall += correct / num_ans
        long_precision += correct / (len(answer['predict'])+1e-5)
        len_long+= 1
    else:
        short_recall += correct / num_ans
        short_precision += correct / (len(answer['predict'])+1e-5)
        len_short += 1

In [11]:
recall /= len(answers)
precision /= len(answers)
long_recall /= len_long
long_precision /= len_long
short_recall /= len_short
short_precision /= len_short
f1 = 2 * recall * precision / (recall + precision + 1e-5)
short_f1 = 2 * short_recall * short_precision / (short_recall + short_precision + 1e-5)
long_f1 = 2 * long_recall * long_precision / (long_recall + long_precision+1e-5)

In [12]:
print(f"Recall: {recall}")
print(f"Precision: {precision}")
print(f"F1: {f1}")
print("=====================================")

print(f"Short Recall: {short_recall}")
print(f"Short Precision: {short_precision}")
print(f"Short F1: {short_f1}")
print("=====================================")

print(f"Long Recall: {long_recall}")
print(f"Long Precision: {long_precision}")
print(f"Long F1: {long_f1}")
print("=====================================")

Recall: 0.4816278166278167
Precision: 0.4387250300532934
F1: 0.4591714650075473
Short Recall: 0.4360215053763441
Short Precision: 0.39011169349691877
Short F1: 0.41178596663490963
Long Recall: 0.6583522727272727
Long Precision: 0.6271017092092451
Long F1: 0.6423421290530603


## Haiku

Recall: 0.4582659932659932 \
Precision: 0.45441863442709507 \
F1: 0.45632920485280953

=====================================

Short Recall: 0.4209677419354839 \
Short Precision: 0.42419216192259934 \
Short F1: 0.4225688012224169

=====================================

Long Recall: 0.6027967171717171 \
Long Precision: 0.5715462153820161 \
Long F1: 0.5867506654834951

=====================================

## Llama 3 8b

Recall: 0.4816278166278167 \
Precision: 0.4387250300532934 \
F1: 0.4591714650075473

=====================================

Short Recall: 0.4360215053763441 \
Short Precision: 0.39011169349691877 \
Short F1: 0.41178596663490963

=====================================

Long Recall: 0.6583522727272727 \
Long Precision: 0.6271017092092451 \
Long F1: 0.6423421290530603 

=====================================

## Llama 3 70b